lib/charms/observability_libs/v1/cert_handler.py

```diff
@@ -0,0 +1,391 @@
+        "Either the library itself is missing (please get it through charmcraft fetch-lib) "
+        "or one of its dependencies is unmet."
+    ) from e
+
+import logging
+
+from ops.charm import CharmBase, RelationBrokenEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import SecretNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+LIBID = "b5cd5cd580f3428fa5f59a8876dcbe6a"
+LIBAPI = 1
+LIBPATCH = 1
+
+
+def is_ip_address(value: str) -> bool:
+    """Return True if the input value is a valid IPv4 address; False otherwise."""
+    try:
+        ipaddress.IPv4Address(value)
+        return True
+    except ipaddress.AddressValueError:
+        return False
+
+
+class CertChanged(EventBase):
+    """Event raised when a cert is changed (becomes available or revoked)."""
+
+
+class CertHandlerEvents(ObjectEvents):
+    """Events for CertHandler."""
+
+    cert_changed = EventSource(CertChanged)
+
+
+class CertHandler(Object):
+    """A wrapper for the requirer side of the TLS Certificates charm library."""
+
+    on = CertHandlerEvents()  # pyright: ignore
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        key: str,
+        certificates_relation_name: str = "certificates",
+        cert_subject: Optional[str] = None,
+        sans: Optional[List[str]] = None,
+    ):
+        """CertHandler is used to wrap TLS Certificates management operations for charms.
+
+        CerHandler manages one single cert.
+
+        Args:
+            charm: The owning charm.
+            key: A manually-crafted, static, unique identifier used by ops to identify events.
+             It shouldn't change between one event to another.
+            certificates_relation_name: Must match metadata.yaml.
+            cert_subject: Custom subject. Name collisions are under the caller's responsibility.
+            sans: DNS names. If none are given, use FQDN.
+        """
+        super().__init__(charm, key)
+
+        self.charm = charm
+        # We need to sanitize the unit name, otherwise route53 complains:
+        # "urn:ietf:params:acme:error:malformed" :: Domain name contains an invalid character
+        self.cert_subject = charm.unit.name.replace("/", "-") if not cert_subject else cert_subject
+
+        # Use fqdn only if no SANs were given, and drop empty/duplicate SANs
+        sans = list(set(filter(None, (sans or [socket.getfqdn()]))))
+        self.sans_ip = list(filter(is_ip_address, sans))
+        self.sans_dns = list(filterfalse(is_ip_address, sans))
+
+        self.certificates_relation_name = certificates_relation_name
+        self.certificates = TLSCertificatesRequiresV2(self.charm, self.certificates_relation_name)
+
+        self.framework.observe(
+            self.charm.on.config_changed,
+            self._on_config_changed,
+        )
+        self.framework.observe(
+            self.charm.on.certificates_relation_joined,  # pyright: ignore
```

Comment:
shouldn't this be `self.charm.on[self.certificates_relation_name].relation_joined`?